src/sync/impl/sync_file.cpp

```diff
@@ -0,0 +1,262 @@
+        if (current == '%') {
+            // Decode. +3.
+            buffer.push_back(decoded_char_for(percent_encoded_string, idx));
+            idx += 3;
+        } else {
+            // No need to decode. +1.
+            if (!character_is_unreserved(current)) {
+                throw std::invalid_argument("Input string is invalid: contains reserved characters.");
+            }
+            buffer.push_back(current);
+            idx++;
+        }
+    }
+    return buffer;
+}
+
+std::string file_path_by_appending_component(const std::string& path, const std::string& component, bool is_directory)
+{
+    // FIXME: Does this have to be changed to accomodate Windows platforms?
+    std::string terminal = "";
+    if (is_directory && component[component.length() - 1] != '/') {
+        terminal = "/";
+    }
+    char path_last = path[path.length() - 1];
+    char component_first = component[0];
+    if (path_last == '/' && component_first == '/') {
+        return path + component.substr(1) + terminal;
+    } else if (path_last == '/' || component_first == '/') {
+        return path + component + terminal;
+    } else {
+        return path + "/" + component + terminal;
+    }
+}
+
+std::string file_path_by_appending_extension(const std::string& path, const std::string& extension)
+{
+    char path_last = path[path.length() - 1];
+    char extension_first = extension[0];
+    if (path_last == '.' && extension_first == '.') {
+        return path + extension.substr(1);
+    } else if (path_last == '.' || extension_first == '.') {
+        return path + extension;
+    } else {
+        return path + "." + extension;
+    }
+}
+
+std::string SyncFileManager::get_utility_directory()
+{
+    auto util_path = file_path_by_appending_component(get_base_sync_directory(), c_utility_directory, true);
+    util::try_make_dir(util_path);
+    return util_path;
+}
+
+std::string SyncFileManager::get_base_sync_directory()
+{
+    auto sync_path = file_path_by_appending_component(m_base_path, c_sync_directory, true);
+    util::try_make_dir(sync_path);
+    return sync_path;
+}
+
+std::string SyncFileManager::user_directory(std::string user_identity)
+{
+    if (user_identity.length() == 0) {
+        throw std::invalid_argument("user_identity cannot be an empty string");
+    }
+    auto user_path = file_path_by_appending_component(get_base_sync_directory(), std::move(user_identity), true);
+    util::try_make_dir(user_path);
+    return user_path;
+}
+
+void SyncFileManager::remove_user_directory(std::string user_identity)
+{
+    if (user_identity.length() == 0) {
+        throw std::invalid_argument("user_identity cannot be an empty string");
+    }
+    auto user_path = file_path_by_appending_component(get_base_sync_directory(), std::move(user_identity), true);
+    try {
+        remove_nonempty_dir(user_path);
+    } catch(File::AccessError) {
+    }
+}
+
+bool SyncFileManager::remove_realm(std::string user_identity, std::string raw_realm_path)
```

Comment:
Right now this method is only used for Realms that the sync manager manages, so not at the moment. If people are interested in using it to remove files generated by older versions of Realm I'll modify the method.